mpas_analysis/shared/generalized_reader/generalized_reader.py

```diff
@@ -0,0 +1,342 @@
+        xarray.DataSet.isel() for details on how this dictonary is used.
+        An example:
+            isel_values = {'nVertLevels': slice(0, 3),
+                           'nCells': cellIDs}
+
+    variable_map : dict, optional
+        A dictionary with keys that are variable names used by
+        MPAS-Analysis and values that are lists of possible names for the same
+        variable in the MPAS dycore that produced the data set (which may
+        differ between versions).
+
+    start_date, end_date : string or datetime.datetime, optional
+        If present, the first and last dates to be used in the data set.  The
+        time variable is sliced to only include dates within this range.
+
+    year_offset : float, optional
+        An offset used to convert an MPAS date to a date in the range supported
+        by xarray (numpy.datetime64).  Resulting dates must be between 1678
+        and 2622.
+
+    Returns
+    -------
+    ds : An xarray data set.
+
+    Raises
+    ------
+    TypeError
+        If the time variable has an unsupported type (not a date string,
+        a floating-pont number of days since the start of the simulation
+        or a numpy.datatime64 object).
+
+    ValueError
+        If the time variable is not found in the data set.
+
+    Author
+    ------
+    Xylar Asay-Davis
+
+    Last modified
+    -------------
+    02/15/2017
+    """
+
+    def preprocess_fn(ds):
+        return _preprocess(ds, calendar=calendar,
+                           time_variable_name=time_variable_name,
+                           variable_list=variable_list,
+                           sel_values=sel_values,
+                           isel_values=isel_values,
+                           variable_map=variable_map,
+                           start_date=start_date,
+                           end_date=end_date,
+                           year_offset=year_offset)
+
+    ds = xarray.open_mfdataset(file_names,
+                               preprocess=preprocess_fn,
+                               decode_times=False, concat_dim='Time')
+
+    ds = mpas_xarray.remove_repeated_time_index(ds)
+
+    if start_date is not None and end_date is not None:
+        if isinstance(start_date, str):
+            start_date = clampToNumpyDatetime64(stringToDatetime(
+                start_date), year_offset)
+        if isinstance(end_date, str):
+            end_date = clampToNumpyDatetime64(stringToDatetime(
+                end_date), year_offset)
+
+    # select only the data in the specified range of dates
+    ds = ds.sel(Time=slice(start_date, end_date))
+
+    return ds  # }}}
+
+
+def _preprocess(ds, calendar, time_variable_name, variable_list, sel_values,
+                isel_values, variable_map, start_date, end_date,
+                year_offset):  # {{{
+    """
+    Overloads the _preprocess function from MpasXarray to include
+    mapping variable names.  See MpasXaray._preprocess for more
+    details.
```

Comment:
@xylar, we aren't overloading anymore, right?  If so, the docstring needs changed.